hypertrace-trace-enricher/trace-reader/src/test/java/org/hypertrace/trace/reader/entities/DefaultTraceEntityReaderTest.java

```diff
@@ -0,0 +1,141 @@
+import org.hypertrace.core.attribute.service.v1.AttributeType;
+import org.hypertrace.core.datamodel.Entity;
+import org.hypertrace.core.datamodel.Event;
+import org.hypertrace.core.datamodel.StructuredTrace;
+import org.hypertrace.entity.data.service.rxclient.EntityDataClient;
+import org.hypertrace.entity.type.service.rxclient.EntityTypeClient;
+import org.hypertrace.entity.type.service.v2.EntityType;
+import org.hypertrace.trace.reader.attributes.TraceAttributeReader;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+@ExtendWith(MockitoExtension.class)
+class DefaultTraceEntityReaderTest {
+  private static final String TENANT_ID = "tenant-id";
+  private static final String TEST_ENTITY_TYPE_NAME = "ENTITY_TYPE_1";
+  private static final String TEST_ENTITY_ID_ATTRIBUTE_KEY = "id";
+  private static final String TEST_ENTITY_NAME_ATTRIBUTE_KEY = "name";
+  private static final AttributeMetadata TEST_ENTITY_ID_ATTRIBUTE =
+      AttributeMetadata.newBuilder()
+          .setScopeString(TEST_ENTITY_TYPE_NAME)
+          .setKey(TEST_ENTITY_ID_ATTRIBUTE_KEY)
+          .setType(AttributeType.ATTRIBUTE)
+          .build();
+  private static final AttributeMetadata TEST_ENTITY_NAME_ATTRIBUTE =
+      AttributeMetadata.newBuilder()
+          .setScopeString(TEST_ENTITY_TYPE_NAME)
+          .setKey(TEST_ENTITY_NAME_ATTRIBUTE_KEY)
+          .setType(AttributeType.ATTRIBUTE)
+          .build();
+  private static final EntityType TEST_ENTITY_TYPE =
+      EntityType.newBuilder()
+          .setAttributeScope(TEST_ENTITY_TYPE_NAME)
+          .setName(TEST_ENTITY_TYPE_NAME)
+          .setIdAttributeKey(TEST_ENTITY_ID_ATTRIBUTE_KEY)
+          .setNameAttributeKey(TEST_ENTITY_NAME_ATTRIBUTE_KEY)
+          .build();
+
+  private static final StructuredTrace TEST_TRACE = defaultedStructuredTraceBuilder().build();
+  private static final Event TEST_SPAN = defaultedEventBuilder().setCustomerId(TENANT_ID).build();
+  private static final Entity BASIC_AVRO_ENTITY =
+      Entity.newBuilder()
+          .setEntityType(TEST_ENTITY_TYPE_NAME)
+          .setEntityId("id-value")
+          .setEntityName("name-value")
+          .setCustomerId(TENANT_ID)
+          .build();
+
+  @Mock EntityTypeClient mockTypeClient;
+  @Mock EntityDataClient mockDataClient;
+  @Mock CachingAttributeClient mockAttributeClient;
+  @Mock TraceAttributeReader mockAttributeReader;
+
+  private DefaultTraceEntityReader entityReader;
+
+  @BeforeEach
+  void beforeEach() {
+    this.entityReader =
+        new DefaultTraceEntityReader(
+            this.mockTypeClient,
+            this.mockDataClient,
+            this.mockAttributeClient,
+            this.mockAttributeReader,
+            new AvroEntityConverter(),
+            new AttributeValueConverter());
+  }
+
+  @Test
+  void canReadAnEntity() {
```

Comment:
It was quite difficult for me to read this test. Mostly because all the expectations are set first and then the actual assertions are a small piece of the test. I feel all this mock expectations can be set in private methods that describe the what and not the how. Something like

`whenTypeClientReturnsTheEntityTypeByName()`
`whenAtributeClientReturnsAttributesForScope()`
etc

Reply:
Makes sense, will update